### ISSUE TITLE
Out of memory error when using horovod.run() to train a Keras/TF model ensemble

### ISSUE THREAD

**Issue by pbruneau** [horovod/horovod#2481]
**Environment:**
1. Framework: TensorFlow, Keras
2. Framework version: TF 2.3.0, Keras 2.4.0
3. Horovod version: 0.21.0
4. MPI version: 3.0.0 (but using Gloo)
5. CUDA version: 10.1
6. NCCL version: 
7. Python version: 3.6.3
8. OS and version: CentOS 7
9. GCC version: 7.2.0

I am trying to train an ensemble of Keras models on a HPC node equipped with 2 GPUs using horovod. Specifically, I would like to train each model using horovod.run() on the 2 GPUs, and load and stitch together the resulting models as an ensemble. To illustrate, I adapted the ``tensorflow2_keras_mnist_runner.py`` to train an ensemble of 4 models, instead of a single one, [gist here](https://gist.github.com/pbruneau/74381ebfd559b1c1975f984dc719e580)

The first model trains fine, but during the second training procedure I get a ``Allocator (GPU_0_bfc) ran out of memory`` (I do not paste the rather lengthy complete trace)

The model in ``tensorflow2_keras_mnist_runner.py`` apparently has approx. 1.9G GPU memory footprint (with 16G on the V100 I'm working with), so in theory I presume that what I'm trying to do should be possible. But I am certainly doing something wrong, though I do not see what - are the ``tf.config.experimental`` placed correctly in the context of ``horovod.run``? I tried to put some ``del xxx \\ gc.collect() \\ K.clear_session() \\ tf.compat.v1.reset_default_graph()`` everywhere, but this did led to the same result.

**Comment by tgaddair** [comment]
Hey @pbruneau, how many V100s are you training with?  If there's only one, or if the outer process is also sharing GPU memory, I would suggest capping the amount of VRAM each worker can allocate (including the outer process):

```
for gpu in gpu_devices:
    tf.config.experimental.set_memory_growth(gpu, True)
    tf.config.set_logical_device_configuration(
        gpu,
        [tf.config.LogicalDeviceConfiguration(
            memory_limit=gpu_memory_limit_mb)])
```

**Comment by pbruneau** [comment]
Hi @tgaddair, I'm using the 2 V100 available on a HPC node for training (as reflected by the ``np=2`` parameter [in my call to ``horovod.run``](https://gist.github.com/pbruneau/74381ebfd559b1c1975f984dc719e580#file-tensorflow2_keras_mnist_runner-py-L113)).

I guess the outer process will query GPU memory due to [my calls to ``load_model`` in ``ensemble_func``](https://gist.github.com/pbruneau/74381ebfd559b1c1975f984dc719e580#file-tensorflow2_keras_mnist_runner-py-L117), and layer creation afterwards. This leads me to another question: imagine I want to run ``ensemble_func`` many times (e.g. for Bayesian hyper-parameter optimization) - can you tell me the right way to proceed in order to make sure the GPU memory footprint is fairly stable through time (i.e. that I don't run out of memory in a distant but certain future)?

**Comment by pbruneau** [comment]
OK I think I found my way thanks to your previous comment @tgaddair! [Updated and working gist here](https://gist.github.com/pbruneau/74381ebfd559b1c1975f984dc719e580).

I first tried to update the GPU config in the ``singlemodel_func`` (the function parametrizing calls to ``horovod.run``). By monitoring the GPU memory footprint using ``nvidia-smi``, I saw that the cap was working as intended. However, after ``horovod.run`` returns, it seemed that the main process wanted to eat up all the cake, [probably due to the ``load_model`` here](https://gist.github.com/pbruneau/74381ebfd559b1c1975f984dc719e580#file-tensorflow2_keras_mnist_runner-py-L131). So naturally, at the next iteration, ``singlemodel_func`` runs out out memory.

The solution was to have the GPU parametrization block [also at the level of the main process](https://gist.github.com/pbruneau/74381ebfd559b1c1975f984dc719e580#file-tensorflow2_keras_mnist_runner-py-L24). This way, the main process gracefully shares VRAM resources with the processed triggered by ``horovod.run``. The script now runs ``ensemble_func`` 10 times without apparent problem.

As my problem is solved, I'm closing the issue. Thank you for the support @tgaddair!
